docs/features/qualityofservice.rst

```diff
@@ -1,55 +1,102 @@
-.. [#f2] If something doesn't work or you get stuck, please review current `QoS issues <https://github.com/search?q=repo%3AThreeMammals%2FOcelot+QoS&type=issues>`_ filtering by |QoS_label| label.
-
-.. |QoS_label| image:: https://img.shields.io/badge/-QoS-D3ADAF.svg
-   :target: https://github.com/ThreeMammals/Ocelot/labels/QoS
+Quality of Service
+==================
+
+    Label: `QoS <https://github.com/ThreeMammals/Ocelot/labels/QoS>`_
+
+Ocelot supports one QoS capability at the current time. You can set on a per Route basis if you want to use a circuit breaker when making requests to a downstream service.
+This uses an awesome .NET library called `Polly <https://www.thepollyproject.org/>`_, check them out `in official repo <https://github.com/App-vNext/Polly>`_.
+
+The first thing you need to do if you want to use the :doc:`../features/administration` API is bring in the relevant NuGet `package <https://www.nuget.org/packages/Ocelot.Provider.Polly>`_:
+
+.. code-block:: powershell
+
+    Install-Package Ocelot.Provider.Polly
+
+Then in your ``ConfigureServices`` method to add Polly services we must call the ``AddPolly()`` extension of the ``OcelotBuilder`` being returned by ``AddOcelot()`` [#f1]_ like below:
+
+.. code-block:: csharp
+
+    services.AddOcelot()
+        .AddPolly();
+
+Then add the following section to a Route configuration: 
+
+.. code-block:: json
+
+  "QoSOptions": {
+    "ExceptionsAllowedBeforeBreaking": 3,
+    "DurationOfBreak": 1000,
+    "TimeoutValue": 5000
+  }
+
+- You must set a number equal or greater than ``2`` against **ExceptionsAllowedBeforeBreaking** for this rule to be implemented. [#f2]_
+- **DurationOfBreak** means the circuit breaker will stay open for 1 second after it is tripped.
+- **TimeoutValue** means if a request takes more than 5 seconds, it will automatically be timed out. 
+
+**Important change:** with polly version 8, ExceptionsAllowedBeforeBreaking must be equal to or greater than 2.
+If you don't want to change your settings, you must use Polly v7 like this:
+.. code-block:: csharp
+
+    services.AddOcelot()
+        .AddPollyV7();
+
+
+You can set the **TimeoutValue** in isolation of the **ExceptionsAllowedBeforeBreaking** and **DurationOfBreak** options:
+
+.. code-block:: json
+
+  "QoSOptions": {
+    "TimeoutValue": 5000
+  }
+
+There is no point setting the other two in isolation as they affect each other!
+
+If you do not add a QoS section, QoS will not be used, however Ocelot will default to a **90** seconds timeout on all downstream requests.
+If someone needs this to be configurable, open an issue. [#f2]_
+
+
+Extensibility
+-------------
+If you want to use your ResiliencePipeline Provider, you can use the following syntax:
+.. code-block:: csharp
+
+    services.AddOcelot()
+        .AddPolly< MyProvider >();
```

Comment:
OK This is new instruction on how to use new generic methods with providers.
But we have new methods like `AddPollyV7`. I don't understand 
- should we add them to docs too?
- should we remove `Obsolete` for `AddPollyV7` because they are actually not obsolete?!

Reply:
When users upgrade, they won't know that they're moving to Polly 8, but just in case, I've kept the old calls, suffixed with V7, so that if some of them need to stay on V7 for technical reasons, they can.

Reply:
OK I see your strategy, Ray! You propose to migrate automatically to Polly v8 pipelines. Great! It looks like system hidden upgrade. But we have to inform the community that we did that migration with possible breaking changes in term of behavior, not code incompatibility.

---
>  if some of them need to stay on V7 for technical reasons, they can.

In this case these methods should be a part of the extensions class forever. And they are not actually `Obsolete`.
So, my concern here is in case of auto-migration (in old methods) and having added new V7-methods, we cannot say that V7 methods are obsolete in the code.

@ggnaegi What do you think?
Do you approve auto-migration strategy?

Reply:
Everything is "auto-migration", when we make major changes to the way we manage pools (e.g. https://github.com/ThreeMammals/Ocelot/pull/1824), isn't that auto-migration? Maybe some people didn't want that.
Most of our users just consume packages without worrying about what's inside.

In my case, on the Polly migration, there are 2 possible approaches, the one I've done (which minimizes the effort for our users to use something better), or else, we make a new package: Polly.v8, and the users have the choice.

For me, the important thing is that our users update the packages and that everything compiles and works as before without modification.

Reply:
I have no objections anymore regarding auto-migrations.
Not an issue!